### ISSUE TITLE
error while building with nix

### ISSUE THREAD

**Issue by BrianHicks** [luc-tielen/souffle-haskell#28]
Hello! I'm building souffle-haskell on a Mac with nix (specifically, I'm pulling this repo down with `niv` and sending `(import sources.souffle-haskell { inherit pkgs; }).souffle-haskell` to my Haskell package, since the ghc865 package set has 0.1 of this library for whatever reason.)

Anyway, when I try to do this—or, simplifying, when I clone the repo and run `nix-build -A souffle-haskell`, I get this error:

```
Building library for souffle-haskell-2.0.0..
[1 of 9] Compiling Language.Souffle.Internal.Bindings
[2 of 9] Compiling Language.Souffle.Internal
[3 of 9] Compiling Language.Souffle.Internal.Constraints
[4 of 9] Compiling Language.Souffle.Marshal
[5 of 9] Compiling Language.Souffle.Class
[6 of 9] Compiling Language.Souffle.Interpreted
[7 of 9] Compiling Language.Souffle.Experimental
[8 of 9] Compiling Language.Souffle.Compiled
[9 of 9] Compiling Paths_souffle_haskell
[1 of 9] Compiling Language.Souffle.Internal.Bindings
error: invalid argument '-std=c++17' not allowed with 'C'

<no location info>: error:
    `cc' failed in phase `C Compiler'. (Exit code: 1)
[3 of 9] Compiling Language.Souffle.Internal.Constraints
error: invalid argument '-std=c++17' not allowed with 'C'

<no location info>: error:
    `cc' failed in phase `C Compiler'. (Exit code: 1)
[9 of 9] Compiling Paths_souffle_haskell
error: invalid argument '-std=c++17' not allowed with 'C'

<no location info>: error:
    `cc' failed in phase `C Compiler'. (Exit code: 1)

builder for '/nix/store/h5jwyndz8bfpf1imilfzqclxjih5mvim-souffle-haskell-2.0.0.drv' failed with exit code 1
error: build of '/nix/store/h5jwyndz8bfpf1imilfzqclxjih5mvim-souffle-haskell-2.0.0.drv' failed
```

I've tried to build with the patch from #27 applied locally, and run into similar problems there (although it gets a little further, it then complains about missing `souffle` in the test phase.)

Any idea what could be happening here? I'd love to be able to use this library, but I'm really beating my head against the wall to get the compiler flags all lined up right!

I've attached the full build log from building at 9c943d5c1d6285c9fa1aa146cf4b9e9ea9d538f7, in case it provides any additional insights.

[build.log](https://github.com/luc-tielen/souffle-haskell/files/5172647/build.log)

**Comment by luc-tielen** [comment]
Hi, thanks for the interest in the library!

I've been really busy the past few weeks, so I haven't managed to finish #27 yet.
The compiler flags are probably the most complicated thing of this entire library. 😅

I re-added `cpp-options` back to the package.yaml in 2.0.0 for a TH function I added (and TH does not take `cxx-options` into account for some reason..). 
What you are seeing is that it's forwarding `cpp-options` to C files (which are compiled with gcc/clang) as well.
GHC8.10 has a [new optcxx flag](https://downloads.haskell.org/~ghc/8.10.2/docs/html/users_guide/flags.html#forcing-options-to-particular-phases) to fix this, but I haven't managed to get that work correctly yet.

I'm considering removing the TH function + cpp-options and then making a release 2.0.1 for that..

Regarding the "souffle not found in test phase": can you give some more information on that? Is this with just using plain Nix?
It might fail because you are instructing it to only build `souffle-haskell`, have you tried just `nix-build`? 
In the README I suggest running `nix-shell` which basically does that + open a shell with everything you need (and is also how I work on the code).

**Comment by luc-tielen** [comment]
Digging a little deeper, I figured out `cpp-options` actually do not refer to C++, but to the Haskell preprocessor 🤦 . 

I managed to figure out a fix which removes cpp-options in package.yaml and still have a working `embedProgram` function by enabling C++ flags on a per-module basis. See #29 for more information.

For me `nix-shell` now works in that PR, `nix-build` still fails because it can't find souffle. I suspect this is a separate issue..

**Comment by luc-tielen** [comment]
@BrianHicks Did the PR solve your problem? Can the issue be closed? 

I'm also curious what you intend to use souffle-haskell for.. 🙂

**Comment by BrianHicks** [comment]
It didn't solve the problem for me, no. 😕 Thank you for thinking through it, though!

In this particular case, I'm walking an import graph to figure out which tests are associated with which source files in an Elm project.

For various reasons, though, I ended up using Rust for the project instead. Not really haskell-souffle related; just that I know the domain of the tool pretty well so I wanted to try a new language to get more perspective on the problems Rust is good at solving. I say that to say this: I'm not blocked by this issue!

Thank you again for your work on this library. I'm looking forward to using it in another context in the future!

**Comment by luc-tielen** [comment]
Sorry to hear that, but good to hear that Rust worked out for you!
It's hard to say what the remaining problem is.. but if you want we could have a call about it.
I could then make a fix / update the README to prevent this from happening to other people.

**Comment by luc-tielen** [comment]
No more reaction, so I'm closing this ticket. If you still want this to be fixed, please reopen the ticket.
